File: src/main/java/com/minecraft/moonlake/reflect/accessors/Accessors.java
/*
 * Copyright (C) 2017 The MoonLake Authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package com.minecraft.moonlake.reflect.accessors;

import com.minecraft.moonlake.MoonLakeAPI;
import com.minecraft.moonlake.api.utility.MinecraftBukkitVersion;
import com.minecraft.moonlake.api.utility.MinecraftVersion;
import com.minecraft.moonlake.builder.Builder;
import com.minecraft.moonlake.builder.SingleParamBuilder;
import com.minecraft.moonlake.reflect.ExactReflect;
import com.minecraft.moonlake.reflect.FuzzyReflect;
import com.minecraft.moonlake.validate.Validate;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

/**
 * <h1>Accessors</h1>
 * 访问器类 (函数暂时不提供doc文档)
 *
 * @version 1.0
 * @author Month_Light
 */
public final class Accessors {

    static {
    }

    private Accessors() {
        super();
    }

    public static FieldAccessor[] getFieldAccessorArray(Class<?> instanceClass, Class<?> fieldClass, boolean forceAccess) {
        List<Field> fieldList = FuzzyReflect.fromClass(instanceClass, forceAccess).getFieldListByType(fieldClass);
        FieldAccessor[] fieldAccessors = new FieldAccessor[fieldList.size()];
        for(int i = 0; i < fieldAccessors.length; i++)
            fieldAccessors[i] = getFieldAccessor(fieldList.get(i));
        return fieldAccessors;
    }

    public static FieldAccessor getFieldAccessor(Class<?> instanceClass, Class<?> fieldClass, boolean forceAccess) {
        Field field = FuzzyReflect.fromClass(instanceClass, forceAccess).getFieldByType(null, fieldClass);
        field.setAccessible(forceAccess);
        return new SimpleFieldAccessor(field);
    }

    public static FieldAccessor getFieldAccessor(Class<?> instanceClass, String fieldName, boolean forceAccess) {
        return getFieldAccessor(ExactReflect.fromClass(instanceClass, forceAccess).getField(fieldName));
    }

    public static FieldAccessor getFieldAccessor(Class<?> instanceClass, int index, boolean forceAccess) {
        return getFieldAccessor(ExactReflect.fromClass(instanceClass, forceAccess).getFieldByIndex(index));
    }

    public static FieldAccessor getFieldAccessor(Field field) {
        return getFieldAccessor(field, true);
    }

    public static FieldAccessor getFieldAccessor(Field field, boolean forceAccess) {
        Validate.checkNotNull(field).setAccessible(forceAccess);
        return new SimpleFieldAccessor(field);
    }

    public static MethodAccessor getMethodAccessor(Class<?> instanceClass, String methodName, Class<?>... params) {
        return getMethodAccessor(ExactReflect.fromClass(instanceClass, true).getMethod(methodName, params));
    }

    public static MethodAccessor getMethodAccessor(Method method) {
        return getMethodAccessor(method, true);
    }

    public static MethodAccessor getMethodAccessor(Method method, boolean forceAccess) {
        Validate.checkNotNull(method).setAccessible(forceAccess);
        return new SimpleMethodAccessor(method);
    }

    public static MethodAccessor getMethodAccessorBuilder(Builder<MethodAccessor> paramBuilder) {
        return Validate.checkNotNull(paramBuilder).build();
    }

    public static MethodAccessor getMethodAccessorBuilderMCVer(SingleParamBuilder<MethodAccessor, MinecraftVersion> paramBuilder) {
        return Validate.checkNotNull(paramBuilder).build(MoonLakeAPI.currentMCVersion());
    }

    public static MethodAccessor getMethodAccessorBuilderBukkitVer(SingleParamBuilder<MethodAccessor, MinecraftBukkitVersion> paramBuilder) {
        return Validate.checkNotNull(paramBuilder).build(MoonLakeAPI.currentBukkitVersion());
    }

    public static <T> ConstructorAccessor<T> getConstructorAccessor(Class<T> instanceClass, Class<?>... params) {
        try {
            return getConstructorAccessor(instanceClass.getConstructor(params));
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(String.format("Unable to find constructor %s(%s).", instanceClass, Arrays.toString(params)));
        } catch (SecurityException e) {
            throw new IllegalStateException("Cannot access constructors.", e);
        }
    }

    public static ConstructorAccessor<?> getConstructorAccessorBuilder(Builder<ConstructorAccessor<?>> paramBuilder) {
        return Validate.checkNotNull(paramBuilder).build();
    }

    public static ConstructorAccessor<?> getConstructorAccessorBuilderMCVer(SingleParamBuilder<ConstructorAccessor<?>, MinecraftVersion> paramBuilder) {
        return Validate.checkNotNull(paramBuilder).build(MoonLakeAPI.currentMCVersion());
    }

    public static ConstructorAccessor<?> getConstructorAccessorBuilderBukkitVer(SingleParamBuilder<ConstructorAccessor<?>, MinecraftBukkitVersion> paramBuilder) {
        return Validate.checkNotNull(paramBuilder).build(MoonLakeAPI.currentBukkitVersion());
    }

    public static <T> ConstructorAccessor<T> getConstructorAccessor(Constructor<T> constructor) {
        return new SimpleConstructorAccessor<>(constructor);
    }

    public static FieldAccessor getFieldAccessorOrNull(Class<?> clazz, String fieldName, Class<?> fieldType) {
        try {
            FieldAccessor accessor = getFieldAccessor(clazz, fieldName, true);
            if(fieldType.isAssignableFrom(accessor.getField().getType()))
                return accessor;
            return null;
        } catch (IllegalArgumentException e) {
        }
        return null;
    }

    public static MethodAccessor getMethodAccessorOrNull(Class<?> clazz, String methodName, Class<?>... params) {
        try {
            return getMethodAccessor(clazz, methodName, params);
        } catch (IllegalArgumentException e) {
        }
        return null;
    }

    public static ConstructorAccessor<?> getConstructorAccessorOrNull(Class<?> clazz, Class<?>... params) {
        try {
            return getConstructorAccessor(clazz, params);
        } catch (IllegalArgumentException e) {
        }
        return null;
    }

    public static MethodAccessor getConstantMethodAccessor(Object returnValue, final Method method) {
        return new MethodAccessor() {
            @Override
            public Object invoke(Object instance, Object... params) {
                return returnValue;
            }

            @Override
            public Method getMethod() {
                return method;
            }
        };
    }

    public static FieldAccessor getSynchronizedFieldAccessor(FieldAccessor fieldAccessor) {
        return fieldAccessor instanceof SynchronizedFieldAccessor ? fieldAccessor : new SynchronizedFieldAccessor(fieldAccessor);
    }

    public final static class SynchronizedFieldAccessor implements FieldAccessor {

        private final FieldAccessor accessor;

        public SynchronizedFieldAccessor(FieldAccessor accessor) {
            this.accessor = accessor;
        }

        @Override
        public Object get(Object instance) {
            return this.accessor.get(instance);
        }

        @Override
        public void set(Object instance, Object value) {
            Object lock = this.accessor.get(instance);
            if(lock != null) {
                synchronized (lock) {
                    this.accessor.set(instance, value);
                }
            } else {
                this.accessor.set(instance, value);
            }
        }

        @Override
        public Field getField() {
            return this.accessor.getField();
        }
    }
}

File: API/src/main/kotlin/com/minecraft/moonlake/api/reflect/accessor/Accessors.kt
/*
 * Copyright (C) 2016-Present The MoonLake (mcmoonlake@hotmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.minecraft.moonlake.api.reflect.accessor

import com.minecraft.moonlake.api.exception.MoonLakeException
import com.minecraft.moonlake.api.reflect.ExactReflect
import com.minecraft.moonlake.api.reflect.FuzzyReflect
import com.minecraft.moonlake.api.throwMoonLake
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

object Accessors {

    /** api */

    @JvmStatic
    @JvmName("getAccessorConstructor")
    fun <T> getAccessorConstructor(constructor: Constructor<T>, forceAccess: Boolean = false): AccessorConstructor<T> {
        constructor.isAccessible = forceAccess
        return AccessorConstructorSimple(constructor)
    }

    @JvmStatic
    @JvmName("getAccessorConstructor")
    @Throws(NoSuchMethodException::class, SecurityException::class)
    fun <T> getAccessorConstructor(clazz: Class<T>, forceAccess: Boolean = false, vararg params: Class<*>): AccessorConstructor<T> {
        val constructor = if(forceAccess) clazz.getDeclaredConstructor(*params) else clazz.getConstructor(*params)
        return getAccessorConstructor(constructor, forceAccess)
    }

    @JvmStatic
    @JvmName("getAccessorConstructorOrNull")
    fun <T> getAccessorConstructorOrNull(clazz: Class<T>, forceAccess: Boolean = false, vararg params: Class<*>): AccessorConstructor<T>? = try {
        getAccessorConstructor(clazz, forceAccess, *params)
    } catch(e: Exception) {
        null
    }

    @JvmStatic
    @JvmName("setAccessorConstructor")
    fun <T> setAccessorConstructor(constructor: Constructor<T>, forceAccess: Boolean = false): Constructor<T> {
        if(forceAccess && !constructor.isAccessible)
            constructor.isAccessible = true
        return constructor
    }

    @JvmStatic
    @JvmName("getAccessorMethod")
    fun getAccessorMethod(method: Method, forceAccess: Boolean = false): AccessorMethod {
        method.isAccessible = forceAccess
        return AccessorMethodSimple(method)
    }

    @JvmStatic
    @JvmName("getAccessorMethod")
    @Throws(NoSuchMethodException::class)
    fun getAccessorMethod(clazz: Class<*>, name: String, forceAccess: Boolean = false, vararg params: Class<*>): AccessorMethod
            = getAccessorMethod(ExactReflect.fromClass(clazz, forceAccess).getMethod(name, *params), forceAccess)

    @JvmStatic
    @JvmName("getAccessorMethodOrNull")
    fun getAccessorMethodOrNull(clazz: Class<*>, name: String, forceAccess: Boolean = false, vararg params: Class<*>): AccessorMethod? = try {
        getAccessorMethod(ExactReflect.fromClass(clazz, forceAccess).getMethod(name, *params), forceAccess)
    } catch(e: Exception) {
        null
    }

    @JvmStatic
    @JvmName("getAccessorMethod")
    @Throws(NoSuchMethodException::class)
    fun getAccessorMethod(clazz: Class<*>, returnType: Class<*>, forceAccess: Boolean = false, params: Array<out Class<*>>): AccessorMethod
            = getAccessorMethod(FuzzyReflect.fromClass(clazz, forceAccess).getMethodByParameters(null, returnType, params), forceAccess)

    @JvmStatic
    @JvmName("getAccessorMethodOrNull")
    fun getAccessorMethodOrNull(clazz: Class<*>, returnType: Class<*>, forceAccess: Boolean = false, params: Array<out Class<*>>): AccessorMethod? = try {
        getAccessorMethod(FuzzyReflect.fromClass(clazz, forceAccess).getMethodByParameters(null, returnType, params), forceAccess)
    } catch(e: Exception) {
        null
    }

    @JvmStatic
    @JvmName("setAccessorMethod")
    fun setAccessorMethod(method: Method, forceAccess: Boolean = false): Method {
        if(forceAccess && !method.isAccessible)
            method.isAccessible = true
        return method
    }

    @JvmStatic
    @JvmName("getAccessorField")
    fun getAccessorField(field: Field, forceAccess: Boolean = false): AccessorField {
        field.isAccessible = forceAccess
        return AccessorFieldSimple(field)
    }

    @JvmStatic
    @JvmName("getAccessorField")
    @Throws(NoSuchFieldException::class)
    fun getAccessorField(clazz: Class<*>, type: Class<*>, forceAccess: Boolean = false): AccessorField
            = getAccessorField(FuzzyReflect.fromClass(clazz, forceAccess).getFieldByType(null, type), forceAccess)

    @JvmStatic
    @JvmName("getAccessorFieldOrNull")
    fun getAccessorFieldOrNull(clazz: Class<*>, type: Class<*>, forceAccess: Boolean = false): AccessorField? = try {
        getAccessorField(FuzzyReflect.fromClass(clazz, forceAccess).getFieldByType(null, type), forceAccess)
    } catch(e: Exception) {
        null
    }

    @JvmStatic
    @JvmName("getAccessorField")
    @Throws(NoSuchFieldException::class)
    fun getAccessorField(clazz: Class<*>, name: String, forceAccess: Boolean = false): AccessorField
            = getAccessorField(ExactReflect.fromClass(clazz, forceAccess).getField(name), forceAccess)

    @JvmStatic
    @JvmName("getAccessorFieldOrNull")
    fun getAccessorFieldOrNull(clazz: Class<*>, name: String, forceAccess: Boolean = false): AccessorField? = try {
        getAccessorField(ExactReflect.fromClass(clazz, forceAccess).getField(name), forceAccess)
    } catch(e: Exception) {
        null
    }

    @JvmStatic
    @JvmName("getAccessorField")
    @Throws(IndexOutOfBoundsException::class)
    fun getAccessorField(clazz: Class<*>, index: Int, forceAccess: Boolean = false): AccessorField
            = getAccessorField(FuzzyReflect.fromClass(clazz, forceAccess).getFieldByIndex(index), forceAccess)

    @JvmStatic
    @JvmName("getAccessorFieldOrNull")
    fun getAccessorFieldOrNull(clazz: Class<*>, index: Int, forceAccess: Boolean = false): AccessorField? = try {
        getAccessorField(FuzzyReflect.fromClass(clazz, forceAccess).getFieldByIndex(index), forceAccess)
    } catch(e: Exception) {
        null
    }

    @JvmStatic
    @JvmName("setAccessorField")
    fun setAccessorField(field: Field, forceAccess: Boolean = false): Field {
        if(forceAccess && !field.isAccessible)
            field.isAccessible = true
        return field
    }

    /** inner class */

    private data class AccessorConstructorSimple<T>(private val constructor: Constructor<T>) : AccessorConstructor<T> {
        override fun newInstance(vararg params: Any?): T = try {
            constructor.newInstance(*params)
        } catch(e: Exception) {
            when(e) {
                is IllegalAccessException -> throw MoonLakeException("无法使用构造函数.", e)
                is InvocationTargetException -> throw MoonLakeException("发生内部错误.", e.cause)
                is InstantiationException -> throw MoonLakeException("无法实例化对象.", e)
                else -> e.throwMoonLake()
            }
        }

        override val accessor: Constructor<T>
            get() = constructor
    }

    private data class AccessorMethodSimple(private val method: Method) : AccessorMethod {
        override fun invoke(instance: Any?, vararg params: Any?): Any? = try {
            method.invoke(instance, *params)
        } catch(e: Exception) {
            when(e) {
                is IllegalAccessException -> throw MoonLakeException("无法使用反射器.", e)
                is InvocationTargetException -> throw MoonLakeException("发生内部错误.", e.cause)
                else -> e.throwMoonLake()
            }
        }

        override val accessor: Method
            get() = method
    }

    private data class AccessorFieldSimple(private val field: Field) : AccessorField {
        override fun get(instance: Any?): Any? = try {
            field.get(instance)
        } catch(e: Exception) {
            when(e) {
                is IllegalArgumentException -> throw MoonLakeException("无法读取字段: $field", e)
                is IllegalAccessException -> throw MoonLakeException("无法使用反射器.", e)
                else -> e.throwMoonLake()
            }
        }

        override fun set(instance: Any?, value: Any?) = try {
            field.set(instance, value)
        } catch(e: Exception) {
            when(e) {
                is IllegalArgumentException -> throw MoonLakeException("无法设置字段 $field 的值为: $value", e)
                is IllegalAccessException -> throw MoonLakeException("无法使用反射器.", e)
                else -> e.throwMoonLake()
            }
        }

        override val accessor: Field
            get() = this.field
    }
}
